ProjFS.Mac/PrjFSKext/Memory.hpp

```diff
@@ -18,21 +18,25 @@ void Memory_Free(void* buffer, uint32_t size);
 template <typename T>
 T* Memory_AllocArray(uint32_t arrayLength)
 {
-    size_t allocBytes = arrayLength * sizeof(T);
-    if (allocBytes > UINT32_MAX)
+    uint32_t allocBytes;
+    if (__builtin_umul_overflow(arrayLength, sizeof(T), &allocBytes))
     {
+        // Overflow occurred.
         return nullptr;
     }
-    
-    return static_cast<T*>(Memory_Alloc(static_cast<uint32_t>(allocBytes)));
+
+    return static_cast<T*>(Memory_Alloc(allocBytes));
 }
 
 template <typename T>
 void Memory_FreeArray(T* array, uint32_t arrayLength)
 {
-    size_t arrayBytes = arrayLength * sizeof(T);
-    assert(arrayBytes <= UINT32_MAX);
-    Memory_Free(array, static_cast<uint32_t>(arrayBytes));
+    uint32_t arrayBytes;
+    if (__builtin_umul_overflow(arrayLength, sizeof(T), &arrayBytes))
+    {
+        assert(!"Overflow detected");
```

Comment:
`panic()` or `assertf()` might be more appropriate for explicitly crashing with a message. `assertf()` in particular would allow us to put the values for `arrayLength` and `sizeof(T)` in the panic message, so we have a chance of working out where the bug is coming from.

Reply:
I like this idea, let me revise the PR.